runtime/src/snapshot_utils.rs

```diff
@@ -1287,17 +1567,136 @@ where
+    snapshot_file_path: P,
+    snapshot_version_path: P,
+    account_paths: Vec<PathBuf>,
+) where
+    P: AsRef<Path>,
+{
+    file_sender
+        .send(snapshot_file_path.as_ref().to_path_buf())
+        .unwrap();
+    file_sender
+        .send(snapshot_version_path.as_ref().to_path_buf())
+        .unwrap();
+
+    for account_path in account_paths {
+        for file in fs::read_dir(account_path).unwrap() {
+            let file_path = file.unwrap().path().to_path_buf();
+            file_sender.send(file_path).unwrap();
+        }
+    }
+}
+
+/// Perform the common tasks when deserialize a snapshot.  Handles reading snapshot file, reading the version file,
+/// and then returning those fields plus the rebuilt storage
+//fn unarchive_snapshot<P, Q>(
+fn build_storage_from_snapshot_file<Q>(
+    bank_snapshots_dir: Q,
+    snapshot_file_path: Q,
+    snapshot_version_path: Q,
+    measure_name: &'static str,
+    account_paths: &[PathBuf],
+    next_append_vec_id: Arc<AtomicU32>,
+) -> Result<UnarchivedSnapshot>
+where
+    Q: AsRef<Path>,
+{
+    let (file_sender, file_receiver) = crossbeam_channel::unbounded();
+
+    // The account files found in accounts/ but not used by the snapshot
+    let unused_account_files = Arc::new(Mutex::new(Vec::<PathBuf>::new()));
+
+    // This hashset is created when parsing the snapshot.  Its entries matching the found account
+    // files are deleted.  When the rebuild is done, if there is any leftover, then some account file is
+    // missing.  The process of constructing from the snapshot directory should be aborted.  It will fall
+    // back to the unarchiving case.
+    let snapshot_appendvecs = Arc::new(Mutex::new(HashSet::<(Slot, AppendVecId)>::new()));
+
+    streaming_snapshot_and_account_files(
+        file_sender,
+        &snapshot_file_path.as_ref(),
+        &snapshot_version_path.as_ref(),
+        account_paths.to_vec(),
+    );
+
+    let num_rebuilder_threads = num_cpus::get_physical().saturating_sub(1).max(1);
+    let (version_and_storages, measure_deserialize) = measure!(
+        SnapshotStorageRebuilder::rebuild_storage(
+            file_receiver,
+            num_rebuilder_threads,
+            next_append_vec_id,
+            SnapshotFrom::File,
+            Some(unused_account_files.clone()),
+            Some(snapshot_appendvecs.clone()),
+        )?,
+        measure_name
+    );
+    info!("{}", measure_deserialize);
+
+    let set = snapshot_appendvecs.lock().unwrap();
+
+    if !set.is_empty() {
+        // Not all snapshot appendvec entries have been matched with the accounts files.  So, the accounts/ files
+        // are not consistent with the snapshot state.  Abort with the error.
+        error!(
+            "Error: Snapshot appendvec entries missing matching accounts files: {:?}",
+            set
+        );
+        return Err(SnapshotError::MissingAccountFile);
+    }
+
+    let vec = unused_account_files.lock().unwrap();
```

Comment:
this will certainly break for incremental snapshots as currently written.

I think we need to not have a multi-part loading where we load full then incremental. We need to get all appendvecs from full and incremental, then run this.

Reply:
MissingAccountFile has been removed.